issue100 Add support for asynchronous checks

xstefank opened the pull request:
resolves #100

cescoffier commented on api/src/main/java/io/smallrye/health/api/AsyncHealthCheck.java:
```diff
@@ -0,0 +1,23 @@
+public interface AsyncHealthCheck {
+
+    /**
+     * Invokes the health check procedure provided by the implementation of this interface.
```
```suggestion
     * Invokes the health check procedure provided by the implementation of this interface.
     * Unlike synchronous checks, this method is used for async check. The returned {@code Uni}
     * propagates the {@code HealthCheckResponse} as item. If the returned {@code Uni} produces
     * a failure, the check is considered as failed.
```

cescoffier commented on api/src/main/java/io/smallrye/health/api/AsyncHealthCheck.java:
```diff
@@ -0,0 +1,23 @@
+
+/**
+ * The async health check procedure interface.
+ * Invoked by consumers to verify the healthiness of a computing node in the reactive manner.
```
```suggestion
 * Invoked by consumers to verify the healthiness of a computing node in an asynchronous manner.
 * This type of check is particularly useful for reactive API inherently asynchronous. 
```

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -148,6 +179,54 @@ private final SmallRyeHealth getHealth(Iterable<HealthCheck>... checks) {
+    }
+
+    @SafeVarargs
+    private final Uni<SmallRyeHealth> getHealthAsync(Iterable<AsyncHealthCheck>... checks) {
```
Suggestion:

The produced Uni can be reused. Every time someone subscribes to that Uni, the checks will be called again.

xstefank replied:
@cescoffier I don't understand your comment. Should I add this to JavaDoc or handle this somehow in the code?

cescoffier replied:
Once the Uni is created, you don't need to re-create it every time. Re-subscribing re-run it. 

```
AtomicInteger count = new AtomicInteger();
Uni<Integer> uni = Uni.createFrom().item(() -> count.getAndIncrement());

uni.await().indefinitely() -> 0
uni.await().indefinitely() -> 1
uni.await().indefinitely() -> 2
uni.await().indefinitely() -> 3
//... 
```

xstefank replied:
ah, so you're suggesting to cache created Uni. But that won't work with dynamic registration as `additionalAsyncChecks` list.

cescoffier replied:
Yes, I realized that... Maybe we can have a cached uni for the "regular" case.

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -133,6 +149,21 @@ public SmallRyeHealth getHealthGroups() {
     }
 
+    @Experimental("Asynchronous Health Check procedures")
+    public Uni<SmallRyeHealth> getHealthAsync() {
```
Suggestion: Instead of distinguishing the async and sync case, you can make sync look async and have a single method:
```
Uni<HealthCheckResponse> syncMadeAsync = Uni.createFrom(() -> sync.call());
```

Note the usage of a supplier which indicates that `sync.call()` is called on subscription (and on every subscription).

xstefank replied:
Can't reuse the sync health calls here as async works with different interface. But this can be useful in tests. Thanks.

cescoffier replied:
That's more something for https://github.com/smallrye/smallrye-health/blob/master/implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java#L137. 

Just have 1 getHealth method instead of 2 and consider the synchronous checks, async.

cescoffier commented on testsuite/experimental/src/test/java/io/smallrye/health/deployment/FailedLivenessAsync.java:
```diff
@@ -0,0 +1,19 @@
+
+    @Override
+    public Uni<HealthCheckResponse> call() {
+        return Uni.createFrom().item(HealthCheckResponse.down(FailedLivenessAsync.class.getName()));
```
To be sure there is no synchronous assumption, add something like:

```
.onItem().delayIt().by(Duration.ofMillis(10)); 
```

cescoffier replied:
It might be good to do that in the other tests.

cescoffier commented on testsuite/experimental/src/test/java/io/smallrye/health/test/AsyncFailedHealthTest.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Test
+    public void testAsyncLiveness() {
+        JsonObject json = smallRyeHealthReporter.getLivenessAsync().await().atMost(Duration.ofSeconds(5)).getPayload();
```
Does it also test the combination of sync and async?

xstefank replied:
I wasn't sure whether to add async checks to sync calls. But it seems it is required.

kenfinnigan commented:
Right now it appears to be the case that sync and async health checks are two parallel streams or options.

As @cescoffier suggested, why not have a single path? When @cescoffier and I had discussed this type of work, for many things, the idea was always to have the reactive pieces as the lowest layer and then build the sync/blocking bits on top.

Even if it means breaking current MP Health APIs, we should consider an approach where it's not two separate processing paths

xstefank commented:
@kenfinnigan already working on it. Should not break anything :)

xstefank commented:
@cescoffier it's still missing Uni caching. I can add that tomorrow afternoon.

cescoffier commented on api/src/main/java/io/smallrye/health/api/AsyncHealthCheck.java:
```diff
@@ -0,0 +1,26 @@
+
+    /**
+     * Invokes the health check procedure provided by the implementation of this interface.
+     * Unlike synchronous checks, this method is used for async check. The returned {@code Uni}
```
```suggestion
     * Unlike synchronous checks, this method is used for asynchronous checks. The returned {@code Uni}
```

cescoffier commented on api/src/main/java/io/smallrye/health/api/AsyncHealthCheck.java:
```diff
@@ -0,0 +1,26 @@
+     * Invokes the health check procedure provided by the implementation of this interface.
+     * Unlike synchronous checks, this method is used for async check. The returned {@code Uni}
+     * propagates the {@code HealthCheckResponse} as item. If the returned {@code Uni} produces
+     * a failure, the check is considered as failed.
```
```suggestion
     * a failure, the check is considered as failed. Returning `null` is invalid and considers the check failed. In addition, returning a `Uni` propagating a `null` value as the item is considered also as a failure.
```

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -69,9 +95,13 @@
     String emptyChecksOutcome;
 
-    private List<HealthCheck> additionalChecks = new ArrayList<>();
+    private final List<HealthCheck> additionalChecks = new ArrayList<>();
```
You could have a single list if every time you add a synchronous additional checks you wrap is as an async check.

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -99,57 +129,169 @@ public void reportHealth(OutputStream out, SmallRyeHealth health) {
 
     public SmallRyeHealth getHealthGroup(String groupName) {
-        return getHealth(allHealthChecks.select(HealthGroup.Literal.of(groupName)));
+        return getHealthGroupAsync(groupName).await().indefinitely();
```
I would use `atMost` and set a timeout. We never know...

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -99,57 +129,169 @@ public void reportHealth(OutputStream out, SmallRyeHealth health) {
-
     public SmallRyeHealth getHealthGroups() {
-        Iterator<Bean<?>> iterator = beanManager.getBeans(HealthCheck.class, Any.Literal.INSTANCE).iterator();
+        return getHealthGroupsAsync().await().indefinitely();
```
I would use `atMost` and set a timeout. We never know...

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -99,57 +129,169 @@ public void reportHealth(OutputStream out, SmallRyeHealth health) {
-        JsonArrayBuilder results = jsonProvider.createArrayBuilder();
-        HealthCheckResponse.State status = HealthCheckResponse.State.UP;
+    private SmallRyeHealth getHealth(HealthType... types) {
+        return getHealthAsync(types).await().indefinitely();
```
I would use atMost and set a timeout. We never know...

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -99,57 +129,169 @@ public void reportHealth(OutputStream out, SmallRyeHealth health) {
+    }
+
+    private Uni<SmallRyeHealth> getHealthAsync(HealthType... types) {
+        List<Iterable<HealthCheck>> checks = new ArrayList<>();
```
You can have a single list.

xstefank replied:
well, I guess I can preprocess CDI instances to get only one list of Unis going downwards.

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -99,57 +129,169 @@ public void reportHealth(OutputStream out, SmallRyeHealth health) {
+    }
+
+    private Uni<SmallRyeHealth> getHealthAsync(Iterable<Iterable<HealthCheck>> checks,
+            Iterable<Iterable<AsyncHealthCheck>> asyncChecks) {
```
You can have a single list containing only async checks.

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -99,57 +129,169 @@ public void reportHealth(OutputStream out, SmallRyeHealth health) {
             for (Iterable<HealthCheck> instance : checks) {
-                status = processChecks(instance, results, status);
+                if (instance == null) {
+                    continue;
```
is this legal? I would say that it should never be `null`.

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -158,57 +300,58 @@ private final SmallRyeHealth getHealth(Iterable<HealthCheck>... checks) {
 
-    private JsonObject jsonObject(HealthCheck check) {
+    private Uni<HealthCheckResponse> callAsync(AsyncHealthCheck asyncHealthCheck) {
+        return withRecovery(asyncHealthCheck.getClass().getName(), asyncHealthCheck.call());
```
You probably want to add a timeout too.

cescoffier replied:
You may want to also handle `null` as item (it should be marked as fail I believe).

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -158,57 +300,58 @@ private final SmallRyeHealth getHealth(Iterable<HealthCheck>... checks) {
+
+    private Uni<HealthCheckResponse> callSync(HealthCheck healthCheck) {
+        return withRecovery(healthCheck.getClass().getName(),
+                Uni.createFrom().item(() -> call(healthCheck)));
```
no need for `call`, it will be a failure.

cescoffier commented:
WARNING: Philosophical question ahead.

When using a `Uni` 3 things can _happen_:

* you get the response (up or down) - easy, you use it
* you don't get a response before a specified timeout - I believe it should fail.
* you get `null` as response - should it fail?

There is also the case where the method returning the `Uni` returns `null` - I would say fail too.

xstefank commented:
@cescoffier I addressed all your comments also now handling null values. Can you take another look, please? Particularly at `callAsync` as I am unable to find a way to handle exceptions from supplier returning null in the Uni itself.

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -69,9 +99,40 @@
 
-    private List<HealthCheck> additionalChecks = new ArrayList<>();
+    @Inject
+    @ConfigProperty(name = "io.smallrye.health.timeout.seconds", defaultValue = "300")
```
300 seconds? That's 5 minutes... isn't this a bit too high?

cescoffier replied:
Shouldn't this property be documented somewhere? (I don't know how you manage doc).

xstefank replied:
We don't have documentation yet but all properties will be there. What would be the best default for timeout? I'll change it to one minute for now.

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -99,57 +160,148 @@ public void reportHealth(OutputStream out, SmallRyeHealth health) {
-        List<HealthCheck> groupHealthChecks = new ArrayList<>();
+    @Experimental("Asynchronous Health Check procedures")
+    public Uni<SmallRyeHealth> getLivenessAsync() {
+        return getHealthAsync(smallRyeLivenessUni, LIVENESS);
```
Are `getLivenessAsync` and `getReadinessAsync` still used (outside of this class)?

xstefank replied:
They are a public experimental API. I would expect that they can be integrated in user applications.

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -158,57 +310,65 @@ private final SmallRyeHealth getHealth(Iterable<HealthCheck>... checks) {
+                        }
+                        return call;
+                    } catch (IllegalArgumentException ex) {
+                        if (ex.getMessage().matches(".*supplier.*null.*")) {
```
Which exception are you trying to catch here?

xstefank replied:
java.lang.IllegalArgumentException: `supplier` must not be `null`

xstefank replied:
I added a tests with the following checks - https://github.com/smallrye/smallrye-health/pull/111/files#diff-0e8e672ac11f35e6ee999ca9f6eeb5d2R79-R98. I don't like this exception handling but I can't find a way how to propagate supplier null exception through the uni itself.

cescoffier replied:
I would just rethrow the exception. 

That exception means that the Uni creation is invalid (`Uni.createFrom().item((Supplier) null);`).

xstefank replied:
ok, that will be easier from the handling perspective. I wanted to have null handling unified but we can go back to this handling if requested by users.

cescoffier commented on implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java:
```diff
@@ -158,57 +310,65 @@ private final SmallRyeHealth getHealth(Iterable<HealthCheck>... checks) {
+    }
+
+    private Uni<HealthCheckResponse> callSync(HealthCheck healthCheck) {
+        // must be a supplier to be reinvoked on every subcribtion, don't change to method reference
```
it will work the same way with a method reference. What should be avoided is: `Uni.createFrom().item(healthCheck.call());`

xstefank replied:
yes, this is exactly what my IDE was trying to change :) thus the comment.

cescoffier replied:
Uni.createFrom().item(healthCheck.call()); != Uni.createFrom().item(HealthCheck::call);

The second one would use the `item` method using the supplier.

xstefank replied:
I'll be dammed. Surely something I've overlooked coding before sleep :)

xstefank commented:
@cescoffier fixed timeout to 1 minute and the method reference.

xstefank commented:
@cescoffier I removed supplier null handling, way nicer now. Can you give it one last green light please?
